packages/engine/paima-funnel/src/funnels/carp/funnel.ts

```diff
@@ -122,11 +156,34 @@ export class CarpFunnel extends BaseFunnel implements ChainFunnel {
       this.sharedData.extensions,
       lastTimestamp,
       this.cache,
-      this.confirmationDepth
+      this.confirmationDepth,
+      this.era
     );
 
     const composed = composeChainData(this.bufferedData, grouped);
 
+    for (const data of composed) {
+      if (!data.internalEvents) {
+        data.internalEvents = [] as InternalEvent[];
+
+        const epoch = absoluteSlotToEpoch(
+          this.era,
+          timestampToAbsoluteSlot(this.era, data.timestamp, this.confirmationDepth)
+        );
+
+        const prevEpoch = this.cache.getState().epoch;
+
+        if (!prevEpoch || epoch !== prevEpoch) {
+          data.internalEvents?.push({
+            type: InternalEventType.CardanoBestEpoch,
+            epoch: epoch,
+          });
+        }
+
+        this.cache.updateEpoch(epoch);
```

Comment:
do we want to update the cache like this for every iteration of this loop? (genuine question)

Reply:
Not really, but afaik the cache it's just another object in memory, so I'm not sure why that would be a problem. Otherwise I could change it to a local variable and then just update it at the end.

Or I can wrap the update in the `epoch !== prevEpoch` condition.

Reply:
yeah it's less about performance and more about understanding when the cache is meant to be updated. If we only care about updating it when `epoch !== prevEpoch` we should put it in the same condition so it's clear when this cache is mean to be updated at a glance